File: json_editor.ts
import {div, empty} from "core/dom"
import * as p from "core/properties"
// import {LayoutDOM, LayoutDOMView} from "models/layouts/layout_dom"
import {InputWidget, InputWidgetView} from "models/widgets/input_widget"
export class JsonEditorView extends InputWidgetView {

  initialize(options) {
    super.initialize(options)

    this.render()

    // Set BokehJS listener so that when the Bokeh slider has a change
    // event, we can process the new data
    this.connect(this.model.change, () => this.render())
  }

  render() {
    // BokehjS Views create <div> elements by default, accessible as
    // ``this.el``. Many Bokeh views ignore this default <div>, and instead
    // do things like draw to the HTML canvas. In this case though, we change
    // the contents of the <div>, based on the current slider value.
    empty(this.el)
    var json = this.model.json
    // container = document.getElementById("jsoneditor")
    var options = {mode: 'view'}
    var container = div({class: "jsoneditor-parent", id:"jsoneditor",})
    var editor = new JSONEditor(container, options)
    editor.set(json)
    this.el.appendChild(container)
    
  }
}

export class JsonEditor extends InputWidget {

  // If there is an associated view, this is typically boilerplate.
  default_view = JsonEditorView

  // The ``type`` class attribute should generally match exactly the name
  // of the corresponding Python class.
  type = "JsonEditor"
}

// The @define block adds corresponding "properties" to the JS model. These
// should normally line up 1-1 with the Python model class. Most property
// types have counterparts, e.g. bokeh.core.properties.String will be
// ``p.String`` in the JS implementation. Any time the JS type system is not
// yet as complete, you can use ``p.Any`` as a "wildcard" property type.
JsonEditor.define({
  json:   [ p.Any ],
})

File: json_editor.py
from bokeh.core.properties import Float, Instance, Tuple, Bool, Enum, String, Any
from bokeh.models import InputWidget
from bokeh.models.callbacks import Callback
from bokeh.core.enums import SliderCallbackPolicy

from bokeh.layouts import column
from bokeh.models import Slider, CustomJS, ColumnDataSource
from bokeh.io import show
from bokeh.plotting import Figure


class JsonEditor(InputWidget):
    # The special class attribute ``__implementation__`` should contain a string
    # of JavaScript (or CoffeeScript) code that implements the JavaScript side
    # of the custom extension model or a string name of a JavaScript (or
    # CoffeeScript) file with the implementation.

    __implementation__ = 'json_editor.ts'
    __javascript__ = ["straxui/static/jsoneditor.min.js"]
    __css__ = ["straxui/static/jsoneditor.min.css"]

    # Below are all the "properties" for this model. Bokeh properties are
    # class attributes that define the fields (and their types) that can be
    # communicated automatically between Python and the browser. Properties
    # also support type validation. More information about properties in
    # can be found here:
    #
    #    https://bokeh.pydata.org/en/latest/docs/reference/core.html#bokeh-core-properties

    disable = Bool(default=True, help="""
    Enable or disable the editor.
    """)

    json = Any(default={}, help="""
    The actual json string to edit.
    """)

    
